Add test for response headers

christian-bromann opened the pull request:
The WebDriver spec defines the response header in the [processing modal section](https://www.w3.org/TR/webdriver1/#processing-model) as follows:

> 3. Set the response’s header with name and value with the following values:
>
> Content-Type
>    "application/json; charset=utf-8"
> Cache-Control
>    "no-cache"

This patch adds tests for this case.

__Question__: the RFC defines header names as case insensitive. Should be lowercase header keys and values by default?

__Note:__: Geckodriver returns a wrong response value for content-type, so I created a bug here: https://bugzilla.mozilla.org/show_bug.cgi?id=1507428

christian-bromann committed:
add test for response headers

jgraham commented on webdriver/tests/new_session/headers.py:
```diff
@@ -0,0 +1,6 @@
+def test_missing_first_match(new_session, add_browser_capabilities):
+    response, _ = new_session({"capabilities": {"alwaysMatch": add_browser_capabilities({})}})
+    assert 'cache-control' in response.headers
```
I think you should lowercase the recieved values here.

christian-bromann replied:
@jgraham see PR question: shall we by default lowercase key and value pairs since headers are case insensitive?

jgraham commented on webdriver/tests/new_session/headers.py:
```diff
@@ -0,0 +1,6 @@
+def test_missing_first_match(new_session, add_browser_capabilities):
```
This isn't really what this is testing. You could possibly even just use the `status` endpoint to test this since it doesn't require a session.

christian-bromann replied:
oh dang, copy paste issue

christian-bromann committed:
change test name

christian-bromann committed:
move header assertion to send_command method

christian-bromann commented:
@jgraham given the fact that this test should be applied to every response I moved the assertion into the `Session` class. I also converted header key/value pairs into lowercase per default.

andreastt commented:
There also appears to be some test failures after one of the recent fixups:
https://travis-ci.org/web-platform-tests/wpt/builds/455470716?utm_source=github_status&utm_medium=notification

christian-bromann committed:
transform headers on assertion

christian-bromann commented:
@andreastt addressed outstanding issues

andreastt commented:
Retriggered some of the jobs and dismissed @jgraham’s old review.

christian-bromann commented:
Azure Pipeline is failing due to
- Chrome not having `cache-control` in command response header
- FF Nightly returning `application/json; charset=utf8` in response header

continuous-integration/travis-ci/pr failing due to
- FF Nightly returning `application/json; charset=utf8` in response header

andreastt commented:
@jgraham Aren’t those jobs supposed to run stability tests, like Travis?

andreastt commented:
Or we could keep the checks in the client and emit a warning when the headers do not conform. I support the idea of moving the hard assertions to a separate test.

andreastt reviewed (changes requested):
See comment above.

christian-bromann committed:
moved assertion to webdriver test harness

christian-bromann commented:
@andreastt @jugglinmike I moved the assertions into the `assert_error` and `assert_success` assertion helpers which are only used in the webdriver test harness. What do you think?

jugglinmike reviewed (approved):
Good idea! Unlike the alternatives we've been considering, this may allow some responses to go untested. Those cases seem rare, so I think the simplicity of this approach wins out.

As for the timeout reported by Taskcluster: [unlike with the previous solution](https://github.com/web-platform-tests/wpt/pull/14074#issuecomment-439517580), the system is now capable of identifying the affected tests:

>     2018-11-19 12:17:10,747 - tc-run - INFO - Identifying tests affected in range 'HEAD^'...
>     2018-11-19 12:17:26,483 - tc-run - INFO - Identified 138 affected tests

That's a lot of tests, possibly enough to exceed our 2-hour limit, though the logs for Chrome are suspiciously short...

andreastt commented:
It looks like both the Chrome and Firefox stability tests timed out after two hours, whereas the regular test results passed. Is the timeout here too short when lots of tests are affected, @jgraham?

whimboo commented:
@jgraham mind having a look at this issue? You may have missed @andreastt's question from Nov 20th. Thanks.

andreastt commented:
Just to be sure, let’s retrigger the tests once more.

whimboo commented:
There are still various job failures for TaskCluster, TravisCI, and Azure Pipelines.

gsnedders commented:
Most of these look like all the brokenness we temporarily had on Friday… :/

whimboo commented:
The newly added assertion has failures:
https://tools.taskcluster.net/groups/cZgXyDz3T0S0m1Sh4X024w/tasks/Ub-qkzjxRoCtmPoz6V_PNg/runs/1/logs/public%2Flogs%2Flive.log#L96762

> TypeError: assert_response_headers() takes exactly 2 arguments (1 given)

@christian-bromann can you please fix that? Thanks

christian-bromann committed:
fixed assert_response_headers method

christian-bromann committed:
add back cache-control check

christian-bromann commented:
@whimboo done

- FF still fails because content type is `application/json; charset=utf8` instead of `application/json; charset=utf-8`
- Chrome still fails as it doesn't have a cache control header

whimboo commented:
I don't know why it fails with Firefox given that we shipped the fix for `utf-8` with geckodriver 0.24.0. @jugglinmike do we still have to manually change the version number of the driver?

jugglinmike commented:
[Nope](https://github.com/web-platform-tests/wpt/commit/480fd85811ed2e3449d48bc0c8b8ec09d0e20412)

whimboo commented:
Oh wait. The Firefox job with the patch applied doesn't fail! What's failing is the verify job, which is killed by taskcluster after a runtime of 7200s.

christian-bromann commented:
@whimboo would closing and reopening the PR help?

whimboo commented:
Not sure. Running all the tests simply take that long. Someone might take a bit of time to check how long those verify tasks run in general, and maybe we have to bump up the maximum allowed time.

gsnedders commented:
@whimboo I'm not sure we can while we're using the GitHub Taskcluster integration, but @jgraham would know better. There are a few cases where we end up running lots of tests, but I think the longer term plan is to do something smarter and dynamically shard them.

In either case, if you think this looks good, we should just force-merge this.

mustjab commented:
@jgraham I see WPT WebDriver tests failing in Chrome on Windows and Linux runs with this change, and pass after reverting it locally. Anyone else seeing this?

```
  Γöé webdriver\tests\add_cookie\add.py:17:
  Γöé _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  Γöé webdriver\tests\support\asserts.py:55: in assert_error
  Γöé     assert_response_headers(response.headers)
  Γöé _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  Γöé
  Γöé headers = {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöé     def assert_response_headers(headers):
  Γöé         """
  Γöé         Method to assert response headers for WebDriver requests
  Γöé
  Γöé         :param headers: dict with header data
  Γöé         """
  Γöé >       assert 'cache-control' in headers
  Γöé E       AssertionError: assert 'cache-control' in {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöé headers    = {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöö webdriver\tests\support\asserts.py:82: AssertionError
```

andreastt commented:
@mustjab I think that would be expected, since chromedriver doesn’t return the mandated `Cache-Control` header.

See [_send a response_](https://w3c.github.io/webdriver/#dfn-send-a-response) in the spec.

mustjab commented:
@andreastt So should we change the way we run these tests? We run them on a daily basis and at this moment most of the tests fail due to this error.

andreastt commented:
Right, the assertion was added as a step to `assert_success()`, causing most—if not all!—of the wdspec tests to fail because the driver doesn’t return the mandated headers.

In order to avoid numerous workarounds, I think the best fix here is to insert a new line into https://cs.chromium.org/chromium/src/chrome/test/chromedriver/server/http_handler.cc?l=1125 and fix chromedriver. Alternatively we could back this change out temporarily to give you time to fix this.

mustjab commented:
Thank you! I'll outreach to ChromeDriver team to fix this and for now, we'll run with this fix internally.

jgraham commented:
I should have read the patch before merging it. I fully agree that this shouldn't be tested for every test but should be a specific set of tests or otherwise isolated from testing other features.

christian-bromann commented:
@jgraham I will tweak the patch and make it so that it is handled in a single test instead of being part of all assertions.

whimboo commented:
@jgraham, so we want to backout the already landed patch for a clean base? I don't think patching on top of the current patch is that great.

andreastt commented:
If @mustjab is fixing chromedriver, I don’t necessarily see the need to move the assertion to separate test, nor a reason to back this out.

Hexcles commented:
https://chromium-review.googlesource.com/c/chromium/src/+/1501997
